pkg/placement/leadership.go

```diff
@@ -0,0 +1,150 @@
+package placement
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+const barrierWriteTimeout = 2 * time.Minute
+
+// MonitorLeadership is used to monitor if we acquire or lose our role
+// as the leader in the Raft cluster. There is some work the leader is
+// expected to do, so we must react to changes
+//
+// reference: https://github.com/hashicorp/consul/blob/master/agent/consul/leader.go
+func (p *Service) MonitorLeadership(parentCtx context.Context) error {
+	if p.closed.Load() {
+		return errors.New("placement is closed")
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		defer cancel()
+		select {
+		case <-parentCtx.Done():
+		case <-p.closedCh:
+		}
+	}()
+
+	raft, err := p.raftNode.Raft(ctx)
+	if err != nil {
+		return err
+	}
+	var (
+		leaderCh                    = raft.LeaderCh()
+		oneLeaderLoopRun            sync.Once
+		loopNotRunning              = make(chan struct{})
+		leaderCtx, leaderLoopCancel = context.WithCancel(ctx)
+	)
+	close(loopNotRunning)
+
+	defer func() {
+		leaderLoopCancel()
+		<-loopNotRunning
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case isLeader := <-leaderCh:
+			if isLeader {
+				oneLeaderLoopRun.Do(func() {
+					loopNotRunning = make(chan struct{})
+					p.wg.Add(1)
+					go func() {
+						defer p.wg.Done()
+						defer close(loopNotRunning)
+						log.Info("Cluster leadership acquired")
+						p.leaderLoop(leaderCtx)
+						oneLeaderLoopRun = sync.Once{}
+					}()
+				})
+			} else {
+				select {
+				case <-loopNotRunning:
+					log.Error("Attempted to stop leader loop when it was not running")
+					continue
+				default:
+				}
+
+				log.Info("Shutting down leader loop")
+				leaderLoopCancel()
+				select {
+				case <-loopNotRunning:
+				case <-ctx.Done():
+					return nil
+				}
+				log.Info("Cluster leadership lost")
+				leaderCtx, leaderLoopCancel = context.WithCancel(ctx)
+				defer leaderLoopCancel()
+			}
+		}
+	}
+}
+
+func (p *Service) leaderLoop(ctx context.Context) error {
+	// This loop is to ensure the FSM reflects all queued writes by applying Barrier
+	// and completes leadership establishment before becoming a leader.
+	for !p.hasLeadership.Load() {
+		// for earlier stop
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			// nop
+		}
+
+		raft, err := p.raftNode.Raft(ctx)
+		if err != nil {
+			return err
+		}
+
+		barrier := raft.Barrier(barrierWriteTimeout)
+		if err := barrier.Error(); err != nil {
+			log.Error("failed to wait for barrier", "error", err)
```

Comment:
Is this a transient error we can ignore? Should we not return the error here and ultimately close placement- given this could mean we can no longer write to the db (?).

Reply:
Yes, this looks safe, because the loop will just restart. If there's a real problem with the raft node, it will lose its leader status anyway, which will cancel the context and the loop.